### ISSUE TITLE
Request feature ported from upstream KeePass: "Show entries of subgroups" 

### ISSUE THREAD

**Issue by Conan1989** [keepassxreboot/keepassxc#1018]
Request feature ported from upstream KeePass: "Show entries of subgroups"

## Expected Behaviour
screenshot of an example view from up stream keepass https://i.imgur.com/6YdmKiS.png
You'll see the root group "example" is selected, and recursively the entries of subgroups are shown.
This setting is controlled under (Toobar) View > "Show entries of subgroups"

## Current Behaviour
The current behaviour in KeePassXC, only the entries of the directly selected group are shown.

## Possible Solution
not within my skill set, sorry

## Steps to Reproduce (for bugs)
N/A

## Context
Makes viewing a highly sorted (lots of groups and sub-groups) easier.

## Debug Info
KeePassXC - Version 2.2.0
Revision: caa49a8ef3ee28ed478192389b21d61107b3b8e0

Libraries:
- Qt 5.9.1
- libgcrypt 1.8.1

Operating system: Arch Linux
CPU architecture: x86_64
Kernel: linux 4.13.3-1-ARCH

Enabled extensions:
- KeePassHTTP
- Auto-Type
- YubiKey

**Comment by frostasm** [comment]
As soon as I find free time I will try to do this task

**Comment by bjornfor** [comment]
Although I've wanted this feature before, now that I'm reading about how KeeShare shares subgroups (recursively), this feature becomes quite important!

When I'm viewing my current root group, KeepassXC gives me the wrong impression of which passwords are going to be shared.

**Comment by droidmonkey** [comment]
Why do you think that is a "wrong impression"? Sharing a group shares all subgroups and all entries in those subgroups.

**Comment by bjornfor** [comment]
Yes, the way sharing (KeeShare) operates recursively on subgroups makes perfect sense to me. The problem is that KeepassXC does _not_ view subgroups recursively. So the set of passwords _viewed_ vs _exported_ are different.

**Comment by droidmonkey** [comment]
Those are two entirely different functions. To me, what is the point of discrete groups if you dump the entire contents in the view anyway?

**Comment by bjornfor** [comment]
> To me, what is the point of discrete groups if you dump the entire contents in the view anyway?

1. My thinking is that it would only dump the contents of that particular group, and its children, but not any other groups, siblings or anything further up the tree. So that behaviour sounds fine to me. If that output is too much, perhaps the user can flatten the tree a bit (use siblings), instead of using deep nesting? And finally, this recursive behaviour could be optional.

2. I looked at https://i.imgur.com/6YdmKiS.png again and I think, by tweaking the requirements a bit, this can be implemented more easily. What if instead of recursively flattening subgroups into the view, the view would show (sub)groups as a "directory entry". It'll be just like a file browser; you see files (passwords) _and directories_ (subgroups) on each level.

**Comment by dec1** [comment]
> To me, what is the point of discrete groups if you dump the entire contents in the view anyway?

Ideally one could choose to display the entries of subgroups or not (eg via a setting as is possible in keepass "View - Show Entries in Subgroups"). 

I generally like to browse via groups but search recursively eg for the entries I most recently edited:
Show Entries in Subgroups - and sort by modification date

**Comment by droidmonkey** [comment]
Search for `*` to show all entries in your database then you can sort by whatever column you want. You can of course search for additional parameters to narrow the view.

**Comment by dec1** [comment]
> Search for `*` to show all entries in your database then you can sort by whatever column you want. You can of course search for additional parameters to narrow the view. If you have `Search current selected group` enabled then it will only show results for the selected group and children.

Excellent. Thanks. 

It is a bit unintuitive though. I wouldn't normally expect this to be able to increase the number of entries shown  (only filter them).

**Comment by loren-wiq** [comment]
I fully expected to see child nodes appear when navigating the hierarchy of a database in keepassxc. I was a little disappointed that this isn't a feature that can be enabled, such as in keepass2.  Please consider adding this as it supports a valid use-case: it makes navigating the myriad number of logins/credenetials/secrets we catalog much more reasonable as it allows you to 'scope in' whatever tree you're currently dealing with. Such as, customers vs internal vs personal information one might record in a keepass database while using it for work. I have over 500 entries in my keepass2 database @ home, I couldn't imagine the frustration of using that file in keepassxc without the ability to see child nodes. It isn't nearly as easily navigable. 

Thank you for your awesome work, great codebase!

**Comment by droidmonkey** [comment]
No offense, but I'd argue your organization is not correct for your use case. Why do you have groups when you want a flat list? Prepend the title with relevant information as necessary. One level of subgroups should be sufficient if you want to narrow the view. Just my two cents on the matter. 

I also have several hundred entries and simply use the search feature, which is very robust, to narrow my view. With the upcoming improvements to global Auto-Type this also becomes rather moot. Ideally you'll be interacting with the Auto-Type prompt and searching across all database in a flat list. It's very efficient.

**Comment by loren-wiq** [comment]
> No offense, but I'd argue your organization is not correct for your use case. Why do you have groups when you want a flat list? Prepend the title with relevant information as necessary. One level of subgroups should be sufficient if you want to narrow the view. Just my two cents on the matter.
> 
> I also have several hundred entries and simply use the search feature, which is very robust, to narrow my view. With the upcoming improvements to global Auto-Type this also becomes rather moot. Ideally you'll be interacting with the Auto-Type prompt and searching across all database in a flat list. It's very efficient.

It was an admittedly poor example of the kind of division of concerns I have built over the last decade (or so it seems) of massaging organizational structure into my keepass2 database. It's not that keepassxc couldn't handle it (of course it could) it's just that the hierarchy lets me organize (for example) by customer, with multiple sites in folders underneath that. Underneath that I may have different projects or subsections related to organizing things on-premise. Putting the site name, etc into the entry's title is a kluge - yes it would work but it's not an ideal solution vs the capability to use 'proper' nested organization in a hierarchy. No offense taken, and I hope you don't take offense to my use of the word kluge :)

Thanks for your hard work!

**Comment by droidmonkey** [comment]
Love a good kludge, lol. This certainly is on our todo list, but it is somewhat of an antipattern.

**Comment by SharkWipf** [comment]
+1, when I moved from traditional Keepass to Keepassxc, this was, and still is, the feature I miss the most (along with a recursive "collapse all"/"expand all" option).  
There's plenty of cases in which I'd like to, i.e., organize/group entries by service, but want to see them all by customer.  
Or where I'd like to see them as a flat structure most of the time, but want to be able to "dig down" to individual servers/services.  
This was the main thing keeping me from switching to Keepassxc for years, and now that I've finally made the switch it's still the biggest pain point for me.  
(Also, a moment of silence for poor @frostasm who hasn't found a moment of free time since 2017 :sweat_smile:)

**Comment by fasbo** [comment]
Sorely miss this feature. Without this it is difficult to locate recently created or modified entries. E.g when trying to reconcile conflicted DB copies, it's a relatively simple matter of opening both files, select top level, subgroup entries shown, and sort all by last modified.

**Comment by droidmonkey** [comment]
Search for `*` and sort by modified.

**Comment by fasbo** [comment]
> Search for `*` and sort by modified.

Great tip, thank-you.

**Comment by UnknownDEV996** [comment]
> Search for `*` and sort by modified.

Hi, I know the post is very old...😅 but I didn't want to open a new one.
could you add the option to associate a group/folder/entity with an already saved search filter?
so everyone can have a sub-folder system, you will need to create the filter but it's better than nothing 😂.
it would be useful if the filter already had basic functions when you click on the folder, otherwise it would already be inserted in the search bar (which however will require a system to clean the search bar when you out-click the folder)
I know that saved filters are visible and clickable, but it doesn't help with organizing folders/groups/entities

Thank you 😊

**Comment by droidmonkey** [comment]
Save a search like: `g:/path/to/group *`

You can also use regex and wildcards to include multiple top level groups

**Comment by UnknownDEV996** [comment]
> Save a search like: `g:/path/to/group *`
> 
> You can also use regex and wildcards to include multiple top level groups

yeah, okay, but my request is to add this option on entity group:
so when we will edit a group or create a new one i can link the filter salved.

in my opinion the search is much simpler, more intuitive and easier because clicking on the folder is easier and faster than clicking on the filter

ty 😊

**Comment by garner1974** [comment]
I also would like to have this feature. Is it possible to vote for a feature in order to show support for this feature? 
Best regards
Chris

**Comment by danielrausch82** [comment]
Comming from KeePass2 I also miss this feature quite alot.

**Comment by jxggx** [comment]
Why we still don't have this feature in BIG 26. 😭😩

**Comment by mixMugz** [comment]
Any ETA?

**Comment by newhinton** [comment]
As #13443 was closed as a duplicate:

(1)I propose a different way to organize this: 

<img width="1278" height="91" alt="Image" src="https://github.com/user-attachments/assets/1995176f-79db-4bfd-b716-798437d0f4ce" />

Here, the subgroup shows up as a singular "folder", it's entries are missing and not shown.
The subgroup itself can be opened by clicking/double-clicking the folder icon.

(2)This was op's suggestion (or the way original keepass handles this):

<img width="1278"  alt="Image" src="https://i.imgur.com/6YdmKiS.png" />

Every entry below the current subgroup is shown by seperators that denote their group.



Both have advantages and disadvantages:

1: Does not break searching or ordering, but "breaks" with the current overview where only entries are in that list
2: Does not mess with the "only entries in list"-convention, but shows a lot more entries that might not be relevant.


I'd like to give this a try eventually, but not without a ux-discussion beforehand about what the best way forward is with this. I don't have a preference, as long as it is possible to enable/disable this feature. (maybe per group?)

**Comment by droidmonkey** [comment]
I dont understand your proposal, it needs to be described more or with a broader example in the picture. 

What interpret your proposal to be is to show a tree view in both the group and entry views. The entry view shows just the subgroup tree of the current group. Is that correct?

**Comment by newhinton** [comment]
Sure!

First some faked screenshots:


1. Root-Group:

<img width="1836" height="1050" alt="Image" src="https://github.com/user-attachments/assets/9544718f-0ee7-4954-91a0-9370f04f5233" />

2. Group 1:

<img width="1836" height="1050" alt="Image" src="https://github.com/user-attachments/assets/f92481f7-eebd-4f35-b7b9-b89beebb2d13" />

3. Group 2: 

<img width="1836" height="1050" alt="Image" src="https://github.com/user-attachments/assets/a0224499-947d-4879-a3da-698153772231" />

4. Group 3: 

<img width="1836" height="1050" alt="Image" src="https://github.com/user-attachments/assets/8ff46f66-27b2-47e6-bdcf-5d878c4ea425" />

5. Group 4:

<img width="1836" height="1050" alt="Image" src="https://github.com/user-attachments/assets/dbc49fb4-1fbf-4f67-a175-e3c79b8f04c1" />

Screenshots 1 and 3 are the important ones. As you can see from the tree-view, both root and Group 2 contain subgroups AND entries. But only their direct entries are shown, and their subgroups. But _not_ the entries of the subgroups.

This will act more like a traditional explorer, because now you have "files" and "folders" and clicking a "folder" opens it and shows its "files".

(Here: entries and groups are shown, clicking a group will open it and show its entries)

This deviates slightly from the broader overview the original issue proposed, which would look like this:

<img width="1836" height="1050" alt="Image" src="https://github.com/user-attachments/assets/bf5948ad-9a0d-4ca6-adff-96aa5f5df3f8" />


(i used a normal entry with a folder icon to "fake" the ui here, the final result needs a different ux for the non-entry rows)
